doc/mm.rst

```diff
@@ -0,0 +1,101 @@
+=======================
+Nim's Memory Management
+=======================
+
+.. default-role:: code
+.. include:: rstcommon.rst
+
+:Author: Andreas Rumpf
+:Version: |nimversion|
+
+..
+
+
+  "The road to hell is paved with good intentions."
+
+
+Introduction
+============
+
+A memory-management algorithm optimal for every use-case cannot exist.
+Nim provides multiple paradigms for needs ranging from large multi-threaded
+applications, to games, hard-realtime systems and small microcontrollers.
+
+This document describes how the management strategies work;
+How to tune the garbage collectors for your needs, like (soft) `realtime systems`:idx:,
+and how the memory management strategies other than garbage collectors work.
+
+
+Multi-paradigm Memory Management Strategies
+===========================================
+
+.. default-role:: option
+
+To choose the memory management strategy use the `--mm:` switch.
+
+**The recommended switch is `--mm:orc`.**
+
+
+ARC/ORC
+-------
+
+`--mm:arc` is roughly comparable to C++'s memory management with `shared_ptr`. However,
```

Comment:
someone coming from non-c++ (python etc) might not be familiar and will wonder at this first paragraph means - for people that are familiar, a distinctive property of `shared_ptr` above all is that it's susceptible to cycles and uses weak references to work around them - this might lead people that are familiar in the wrong direction.

The presentation of arc is also a bit odd: the first thing the text says about it is that you enable it if you want leaks in async code (which by the way is a bit of a red herring - it's not only async code that doesn't work with arc).

Given that "orc" is recommended, it seems reasonable to describe ARC in terms of what it removes from ORC, and under what circumstances it can actually be expected to work, rather than the other way around.

Reply:
Good points.